### ISSUE TITLE
API Documentation - Show inherited members

### ISSUE THREAD

**Issue by peardox** [castle-engine/castle-engine#177]
If you go to https://castle-engine.io/apidoc-unstable/html/CastleScene.TCastleScene.html you will see a big list of all the APIs exposed by TCastleScene but nowhere on that page will you see any reference for the Load methods from TCastleSceneCore.

It would be beneficial if all APIs (optionally?) listed the inherited methods, properties, events etc as unless you know the API and hierarchy by heart (@michaliskambi  and a very few others being candidates) you may end up missing something already provided. 

For those mere mortals, like me, something like the FPC API documentation makes things far easier to digest. Nice lists of alphabetically or hierarchy sorted entries showing it all.

This could be alternate documentation released in step with stable releases.

**Comment by eugeneloza** [comment]
The API documentation is generated by PasDoc (https://github.com/pasdoc/pasdoc) on the first glance I didn't see any option to generate the API documentation in this way. However, indeed, I remember it got me several months (if not years) to get used to using the online documentation efficiently, and even now I often prefer `ctrl+click` in Lazarus to go to the function definition (and thus comment) than to scan a web version for which ancestor the method or property belongs to. So it might be a useful addition (e.g. as an alternative documentation version - some complex classes API may become simply overwhelming in this case), I also prefer using `ctrl+F` in browser to find things :). I highly doubt it is something easily achievable though.

**Comment by peardox** [comment]
I thought that maybe I was being overly fussy with this one, but apparently not... 

I also often resort to ctrl+click then ctrl+F.

I think that, especially in the case of a newbie, they will get very lost very quickly resulting, in the worst scenario, choosing an alternate solution (Unity gets another customer)

Something else that just occurred to me is that the FPC stuff has, usually, a code snippet for every entry. 

For example I'm currently looking blankly at TX3DRootNode.SaveToStream(Writer: TX3DWriter) vainly grepping thru examples + tools in the hope of possibly saving my TX3DRootNode to a file (I know view3dscene does something like it so looking there now)

**Comment by michaliskambi** [comment]
I agree we should show the inherited members, and it will be done :) PasDoc doesn't have an option to show it yet, but it should be easy to add this feature (I'm also the maintainer of PasDoc now - PRs to PasDoc most welcome. :) ).

I also agree it's important. For a typical user, the difference between whether something is part of TCastleScene or TCastleSceneCore doesn't usually matter, you will almost always use TCastleScene with all features. So you should be able to more easily find Load member on TCastleScene.

P.S. To save the TX3DRootNode you should use Save3D, example on https://github.com/castle-engine/castle-engine/blob/master/examples/3d_rendering_processing/build_3d_object_by_code.lpr . It's not good (and I will see can I hide them) that versions with TX3DWriter arguments confused you, they should be treated as internal. I'll see into hiding them better:)

**Comment by michaliskambi** [comment]
I updated the issue title to better reflect what is needed.

**Comment by peardox** [comment]
Yeah, I knew you were PasDoc maintainer...

A dumb thought, possibly not practical (or possible). In Editor you stream components to file. Now, pick a component deep in the hierarchy - I'll go for TCastleSaveImageDialog. The API listing for this has it descending from TSavePictureDialog - that's it.

However visually examining https://castle-engine.io/apidoc-unstable/html/ClassHierarchy.html I find the following hierarchy...

TObject
TPersistent
TComponent
TLCLComponent
TCommonDialog
TFileDialog
TOpenDialog
TPreviewFileDialog
TOpenPictureDialog
TSavePictureDialog
TCastleSaveImageDialog

Is it not possible with RTTI to obtain all details of the parents and thereby all things available to TCastleSaveImageDialog?

**Comment by michaliskambi** [comment]
It's simply an error that on https://castle-engine.io/apidoc-unstable/html/CastleDialogs.TCastleSaveDialog.html we don't display all ancestors, up to the TObject. We should, we have the necessary information. I'll fix it too:)

Note that RTTI is not related here. PasDoc doesn't and will not (it would require very different preparation of code for docs) compile the code. So it cannot use RTTI. However PasDoc parses the provided code and it knows the hierarchy of external classes through a special feature https://github.com/pasdoc/pasdoc/wiki/ExternalClassHierarchy . CGE is using this, and it provides info about various FPC and LCL classes this way.

(As it happens, this also seems to trigger the bug. In class hierarchy diagram, PasDoc follows the extra "external hierarchy" info. But on particular class page, it stops at first external class. That's just an omission in implementation (mine :) ) and I can easily fix that.)

**Comment by michaliskambi** [comment]
Showing the full ancestors is now fixed in PasDoc, so https://castle-engine.io/apidoc-unstable/html/CastleDialogs.TCastleSaveDialog.html will now show all ancestors.

(The page will actually update only once this change is propagated to Docker image with pasdoc binary that updates API docs. I kicked it manually now, so should be done tomorrow.)

**Comment by michaliskambi** [comment]
To directly address the use-case from the description: our [TCastleScene API docs](https://castle-engine.io/apidoc/html/CastleScene.TCastleScene.html) will show `Load` method docs (from ancestor `TCastleSceneCore`) once you click checkbox "Inherited".
